gobblin-metrics/src/test/java/gobblin/metrics/performance/MetricsPerformanceTest.java

```diff
@@ -0,0 +1,233 @@
+    }
+    if(timers.isEmpty()) {
+      timers = Sets.newHashSet(0);
+    }
+    if(queries == 0) {
+      queries = 50000000l;
+    }
+
+    Set<List<Integer>> parameters = Sets.cartesianProduct(threads, depth, forkAtDepth, counters, meters,
+        histograms, timers);
+
+    Comparator<List<Integer>> comparator = new Comparator<List<Integer>>() {
+      @Override public int compare(List<Integer> o1, List<Integer> o2) {
+        Iterator<Integer> it1 = o1.iterator();
+        Iterator<Integer> it2 = o2.iterator();
+
+        while(it1.hasNext() && it2.hasNext()) {
+          int compare = Integer.compare(it1.next(), it2.next());
+          if(compare != 0) {
+            return compare;
+          }
+        }
+        if(it1.hasNext()) {
+          return 1;
+        } else if(it2.hasNext()) {
+          return -1;
+        } else {
+          return 0;
+        }
+      }
+    };
+
+    TreeMap<List<Integer>, Double> results = Maps.newTreeMap(comparator);
+
+    for(List<Integer> p : parameters) {
+      results.put(p, _singleTest(p.get(0), p.get(1), p.get(2), p.get(3), p.get(4), p.get(5), p.get(6), queries));
+    }
+
+    log.info("Threads\tDepth\tForkAtDepth\tCounters\tMeters\tHistograms\tTimers\tQPS");
+    for(Map.Entry<List<Integer>, Double> result : results.entrySet()) {
+      List<Integer> p = result.getKey();
+      log.info(String.format("%d\t%d\t%d\t%d\t%d\t%d\t%d\t%f", p.get(0), p.get(1), p.get(2), p.get(3), p.get(4),
+          p.get(5), p.get(6), result.getValue()));
+    }
+  }
+
+  /**
+   * Runs a single performance test. Creates a {@link gobblin.metrics.MetricContext} tree, spawns a number of threads,
+   * uses and {@link Incrementer} to update the metrics repeatedly, then determines the achieved QPS in number
+   * of iterations of {@link Incrementer} per second.
+   *
+   * @param threads Number of threads to spawn. Each thread will have an {@link Incrementer} and update metrics.
+   * @param depth Depth of the {@link gobblin.metrics.MetricContext} tree.
+   * @param forkAtDepth If multiple threads, each thread has its own {@link gobblin.metrics.MetricContext}. This
+   *                    parameter sets the first level in the tree where the per-thread MetricContexts branch off.
+   * @param counters Number of counters to generate per thread.
+   * @param meters Number of meters to generate per thread.
+   * @param histograms Number of histograms to generate per thread.
+   * @param timers Number of timers to generate per thread.
+   * @param queries Number of increments to do, divided among all threads.
+   * @return total QPS achieved (e.g. total increments per second in the {@link Incrementer}s)
+   * @throws Exception
+   */
+  @Builder(buildMethodName = "run", builderMethodName = "singleTest")
+  public static double _singleTest(int threads, int depth, int forkAtDepth, int counters, int meters, int histograms,
+      int timers, long queries) throws Exception {
+
+    System.gc();
```

Comment:
I put it there to clean `MetricContext`s from other runs to ideally reduce overhead. It shouldn't affect anything negatively.